lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,768 @@
+    receive do
+      {:DOWN, _ref, :process, pid, reason} ->
+        case pids do
+          %{^pid => child} ->
+            stacks = wait_child(pid, child, reason, stacks)
+            wait_children(pids, size - 1, timers, stacks)
+
+          %{} ->
+            wait_children(pids, size, timers, stacks)
+        end
+
+      {:timeout, timer, :kill} ->
+        for pid <- Map.fetch!(timers, timer), do: Process.exit(pid, :kill)
+        wait_children(pids, size, Map.delete(timers, timer), stacks)
+    end
+  end
+
+  defp wait_child(pid, {_, _, :brutal_kill, _, _} = child, reason, stacks) do
+    case reason do
+      :killed -> stacks
+      _ -> Map.put(stacks, pid, {child, reason})
+    end
+  end
+
+  defp wait_child(pid, {_, restart, _, _, _} = child, reason, stacks) do
+    case reason do
+      :shutdown -> stacks
+      :normal when restart != :permanent -> stacks
+      reason -> Map.put(stacks, pid, {child, reason})
+    end
+  end
+
+  defp maybe_restart_child(pid, reason, %{children: children} = state) do
+    case children do
+      %{^pid => {_, restart, _, _, _} = child} ->
+        maybe_restart_child(restart, reason, pid, child, state)
+
+      %{} ->
+        {:ok, state}
+    end
+  end
+
+  defp maybe_restart_child(:permanent, reason, pid, child, state) do
+    report_error(:child_terminated, reason, pid, child, state)
+    restart_child(pid, child, state)
+  end
+
+  defp maybe_restart_child(_, :normal, pid, _child, state) do
+    {:ok, delete_child(pid, state)}
+  end
+
+  defp maybe_restart_child(_, :shutdown, pid, _child, state) do
+    {:ok, delete_child(pid, state)}
+  end
+
+  defp maybe_restart_child(_, {:shutdown, _}, pid, _child, state) do
+    {:ok, delete_child(pid, state)}
+  end
+
+  defp maybe_restart_child(:transient, reason, pid, child, state) do
+    report_error(:child_terminated, reason, pid, child, state)
+    restart_child(pid, child, state)
+  end
+
+  defp maybe_restart_child(:temporary, reason, pid, child, state) do
+    report_error(:child_terminated, reason, pid, child, state)
+    {:ok, delete_child(pid, state)}
+  end
+
+  defp delete_child(pid, state) do
+    %{children: children, dynamic: dynamic} = state
+    %{state | children: Map.delete(children, pid), dynamic: dynamic - 1}
+  end
+
+  defp restart_child(pid, child, state) do
+    case add_restart(state) do
+      {:ok, %{strategy: strategy} = state} ->
+        case restart_child(strategy, pid, child, state) do
+          {:ok, state} ->
+            {:ok, state}
+
+          {:try_again, state} ->
+            send(self(), {:"$gen_restart", pid})
+            {:ok, state}
+        end
+
+      {:shutdown, state} ->
+        report_error(:shutdown, :reached_max_restart_intensity, pid, child, state)
+        {:shutdown, delete_child(pid, state)}
+    end
+  end
+
+  defp add_restart(state) do
+    %{max_seconds: max_seconds, max_restarts: max_restarts, restarts: restarts} = state
+    now = :erlang.monotonic_time(1)
```

Comment:
`:second` would be clearer I think?

Reply:
I wanted to avoid the second and seconds mess since we support 19.0. I will drop a comment.